pkg/utils/webhook.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/argoproj-labs/applicationset/api/v1alpha1"
+	"github.com/argoproj-labs/applicationset/common"
+	argosettings "github.com/argoproj/argo-cd/v2/util/settings"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/go-playground/webhooks.v5/github"
+	"gopkg.in/go-playground/webhooks.v5/gitlab"
+)
+
+type WebhookHandler struct {
+	namespace string
+	github    *github.Webhook
+	gitlab    *gitlab.Webhook
+	client    client.Client
+}
+
+func NewWebhookHandler(namespace string, argocdSettingsMgr *argosettings.SettingsManager, client client.Client) *WebhookHandler {
+	// register the webhook secrets stored under "argocd-secret" for verifying incoming payloads
+	argocdSettings, err := argocdSettingsMgr.GetSettings()
+	if err != nil {
+		log.Errorf("Failed to get argocd settings: %v", err)
+	}
+	githubHandler, err := github.New(github.Options.Secret(argocdSettings.WebhookGitHubSecret))
+	if err != nil {
+		log.Warnf("Unable to init GitHub webhook: %v", err)
+	}
+	gitlabHandler, err := gitlab.New(gitlab.Options.Secret(argocdSettings.WebhookGitLabSecret))
+	if err != nil {
+		log.Warnf("Unable to init GitLab webhook: %v", err)
+	}
+
+	return &WebhookHandler{
+		namespace: namespace,
+		github:    githubHandler,
+		gitlab:    gitlabHandler,
+		client:    client,
+	}
+}
+
+func (h *WebhookHandler) HandleEvent(payload interface{}) {
+	webURL, revision, touchedHead := getPayloadInfo(payload)
+	log.Infof("Received push event repo: %s, revision: %s, touchedHead: %v", webURL, revision, touchedHead)
+
+	appSetList := &v1alpha1.ApplicationSetList{}
+	err := h.client.List(context.Background(), appSetList, &client.ListOptions{})
+	if err != nil {
+		log.Warnf("Failed to list applicationsets: %v", err)
+		return
+	}
+
+	urlObj, err := url.Parse(webURL)
+	if err != nil {
+		log.Warnf("Failed to parse repoURL '%s'", webURL)
+		return
+	}
+	regexpStr := `(?i)(http://|https://|\w+@|ssh://(\w+@)?)` + urlObj.Hostname() + "(:[0-9]+|)[:/]" + urlObj.Path[1:] + "(\\.git)?"
+	repoRegexp, err := regexp.Compile(regexpStr)
+	if err != nil {
+		log.Warnf("Failed to compile regexp for repoURL '%s'", webURL)
+		return
+	}
+
+	for _, appSet := range appSetList.Items {
+		for _, gen := range appSet.Spec.Generators {
+			// check if the ApplicationSet uses the git generator that is relevant to the payload
+			if gen.Git != nil && gitGeneratorUsesURL(gen.Git, revision, repoRegexp) && genRevisionHasChanged(gen.Git, revision, touchedHead) {
+				err := refreshApplicationSet(h.client, &appSet)
+				if err != nil {
+					log.Warnf("Failed to refresh ApplicationSet '%s' for controller reprocessing", appSet.Name)
+					continue
+				}
+			}
+		}
+	}
+}
+
+func (h *WebhookHandler) Handler(w http.ResponseWriter, r *http.Request) {
+	var payload interface{}
+	var err error
+
+	switch {
+	case r.Header.Get("X-GitHub-Event") != "":
+		payload, err = h.github.Parse(r, github.PushEvent)
```

Comment:
github/gitlab handler could be nil if their New() function have failed?

Reply:
I've updated the `NewGitHubHandler` to return an error. This function won't be called if there's an error while creating handler